src/msg/Message.cc

```diff
@@ -311,9 +311,9 @@ Message *decode_message(CephContext *cct,
                         int crcflags,
                         ceph_msg_header& header,
                         ceph_msg_footer& footer,
-                        ceph::bufferlist& front,
-                        ceph::bufferlist& middle,
-                        ceph::bufferlist& data,
+                        ceph::bufferlist&& front,
+                        ceph::bufferlist&& middle,
+                        const ceph::bufferlist& data,
```

Comment:
`set_data()` callls `share()` on the bufferlist. I haven't checked yet why exactly; hopefully this could replaced with the moving behavior.

Reply:
I don't know enough of the history of Ceph to answer this, but I can do a bit of code archaeology:

This commit added the data.share():

https://github.com/ceph/ceph/commit/28725eb2e4e97f7cf06ffae9445c96df388a6a2e

It looks like its part of a series of 4 commits:
 
https://github.com/ceph/ceph/commits/main/?since=2014-12-11&until=2014-12-11&after=b9d6a0075659fd4810cfaa008dcd10f63ed98593+34

Much of the code for sharable buffers was later removed by:

https://github.com/ceph/ceph/pull/32934

So my guess is the call to share() is no longer needed and this could be change to use move

Reply:
cc @mattbenjamin 

my recollection is that XioMessenger was using registered memory, so we wanted to prevent ceph daemons from holding long-term references to them. so calling share() would copy the buffers into heap memory so the original buffers could be reused

msg/async/rdma also appears to deal with registered memory, so may have similar requirements. unsure about msg/async/dpdk or crimson's messenger